akka-actor/src/main/scala/akka/io/InetAddressDnsResolver.scala

```diff
@@ -1,34 +1,68 @@
+/**
+ * Copyright (C) 2016-2017 Lightbend Inc. <http://www.lightbend.com>
+ */
 package akka.io
 
-import java.net.{ UnknownHostException, InetAddress }
+import java.net.{ InetAddress, UnknownHostException }
+import java.security.Security
 import java.util.concurrent.TimeUnit
 
 import akka.actor.Actor
 import com.typesafe.config.Config
 
 import scala.collection.immutable
-import sun.net.{ InetAddressCachePolicy ⇒ IACP }
 import akka.util.Helpers.Requiring
 
+import scala.util.Try
+
+/** Respects the settings that can be set on the Java runtime via parameters. */
 class InetAddressDnsResolver(cache: SimpleDnsCache, config: Config) extends Actor {
 
-  import IACP.NEVER // 0 constant
+  // Controls the cache policy for successful lookups only
+  private final val CachePolicyProp = "networkaddress.cache.ttl"
+  @deprecated("Deprecated property key, keeping for legacy compatibility; replaced by CachePolicyProp") 
+  private final val CachePolicyPropFallback = "sun.net.inetaddr.ttl"
+
+  // Controls the cache policy for negative lookups only
+  private final val NegativeCachePolicyProp = "networkaddress.cache.negative.ttl"
+  @deprecated("Deprecated property key, keeping for legacy compatibility; replaced by NegativeCachePolicyProp") 
+  private final val NegativeCachePolicyPropFallback = "sun.net.inetaddr.negative.ttl"
+
+  // default values (-1 and 0 are magic numbers, trust them)
+  private final val Forever = -1
+  private final val Never = 0
+  private final val DefaultPositive = 30
+
+  private lazy val cachePolicy: Int = {
+    val n = Try(Security.getProperty(CachePolicyProp).toInt)
+      .orElse(Try(Security.getProperty(CachePolicyPropFallback).toInt))
+      .getOrElse(DefaultPositive) // default
+    if (n < 0) Forever else n
+  }
+
+  private lazy val negativeCachePolicy = {
+    val n = Try(Security.getProperty(NegativeCachePolicyProp).toInt)
+      .orElse(Try(Security.getProperty(NegativeCachePolicyPropFallback).toInt))
+      .getOrElse(0) // default
+    if (n < 0) Forever else n
+  }
 
   private def getTtl(path: String, positive: Boolean): Long =
     config.getString(path) match {
       case "default" ⇒
-        (if (positive) IACP.get else IACP.getNegative) match {
-          case NEVER      ⇒ NEVER
+        (if (positive) cachePolicy else negativeCachePolicy) match {
+          case Never      ⇒ Never
           case n if n > 0 ⇒ TimeUnit.SECONDS.toMillis(n)
           case _          ⇒ Long.MaxValue // forever if negative
         }
       case "forever" ⇒ Long.MaxValue
-      case "never"   ⇒ NEVER
+      case "never"   ⇒ Never
       case _ ⇒ config.getDuration(path, TimeUnit.MILLISECONDS)
         .requiring(_ > 0, s"akka.io.dns.$path must be 'default', 'forever', 'never' or positive duration")
     }
-  val positiveTtl = getTtl("positive-ttl", true)
-  val negativeTtl = getTtl("negative-ttl", false)
+
+  val positiveTtl: Long = getTtl("positive-ttl", positive = true)
```

Comment:
Why is positive public but negative private?

Reply:
both can be private thanks